lib/subscription/subscription-store.ts

```diff
@@ -1,72 +1,47 @@
 /// <reference path='../../typings/tsd.d.ts' />
 
 import _ = require('lodash');
-import Subscription = require('./subscription');
+import Subscription = require('./subscription-buffer');
+import Map = require('../util/map');
+import interfaces = require('../interface');
 
 export = SubscriptionStore;
 
-class SubscriptionStore<T extends Subscription> {
+class SubscriptionStore extends Map<Subscription> {
 
-    protected subscriptions: {[key: number]: T} = {};
-
-    get size(): number {
-        return _.keys(this.subscriptions).length;
-    }
-
-    get(cid: number): T {
-        return this.has(cid) ? this.subscriptions[cid] : undefined;
-    }
-
-    getAll(): T[] {
-        return _.values(this.subscriptions);
-    }
-
-    add(subscription: T): boolean {
-        if (this.has(subscription.correlationId)) {
-            return false;
-        }
-        this.subscriptions[subscription.correlationId] = subscription;
-        return true;
-    }
-
-    has(value: number): boolean;
-    has(value: T): boolean;
-    has(value: any): boolean {
-        if (typeof value === 'number') {
-            return _.has(this.subscriptions, value);
-        }
-        else if (typeof value === 'object') {
-            return _.has(this.subscriptions, value.correlationId);
+    startAllNewBuffers(): Object[] {
+        var buffers: Object[] = [];
+        // Check if we have new data for any subscriptions
+        var hasNewData: boolean = _.some(this._map, (sub: Subscription): boolean => {
+            return !sub.buffer.isEmpty();
+        });
+        if (hasNewData) { // Start a new buffer for every subscription and return back the new data
+            this.forEach((sub: Subscription, key: string, map: Map<Subscription>): void => {
+                var buff: interfaces.IBufferedData<Object> = sub.buffer.startNewBuffer();
+                if (buff.buffer.length) {
+                    buffers.push({
+                        'correlationId': sub.correlationId,
+                        'data': buff.buffer,
+                        'missed_ticks': buff.overflow
+                    });
+                }
+            });
         }
-        return false;
+        return buffers;
     }
 
-    delete(value: number): boolean;
-    delete(value: T): boolean;
-    delete(value: any): boolean {
-        if (typeof value === 'number') {
-            if (!this.has(value)) {
-                return false;
-            }
-            delete this.subscriptions[value];
-            return true;
-        }
-        else if (typeof value === 'object') {
-            if (!this.has(value)) {
-                return false;
+    getAllOldBuffers(): Object[] {
+        var buffers: Object[] = [];
+        this.forEach((sub: Subscription, key: string, map: Map<Subscription>): void => {
```

Comment:
Is there a way to tell `_.map()` not add an element to the array?(ie, the return array's length is less by or equal to the original collection)